Mathlib/RingTheory/Smooth/StandardSmooth.lean

```diff
@@ -0,0 +1,206 @@
+to the coordinates `P.map i` for all `i : P.rels`.
+
+The determinant of this map is the jacobian of `P` used to define when a `PreSubmersivePresentation`
+is submersive. See `PreSubmersivePresentation.jacobian`.
+-/
+noncomputable def differential : (P.rels → P.Ring) →ₗ[P.Ring] (P.rels → P.Ring) :=
+  Basis.constr P.basis P.Ring
+    (fun j i : P.rels ↦ MvPolynomial.pderiv (P.map i) (P.relation j))
+
+/-- The jacobian of a `P : PreSubmersivePresentation` is the determinant
+of `P.linearMap` viewed as element of `S`. -/
+noncomputable def jacobian : S :=
+  algebraMap P.Ring S <| LinearMap.det P.differential
+
+section Matrix
+
+variable [Fintype P.rels] [DecidableEq P.rels]
+
+/--
+If `P.rels` has a `Fintype` and `DecidableEq` instance, the differential of `P`
+can be expressed in matrix form.
+-/
+noncomputable def jacobiMatrix : Matrix P.rels P.rels P.Ring :=
+  LinearMap.toMatrix P.basis P.basis P.differential
+
+lemma jacobian_eq_jacobiMatrix_det : P.jacobian = algebraMap P.Ring S P.jacobiMatrix.det := by
+   simp [jacobiMatrix, jacobian]
+
+lemma jacobiMatrix_apply (i j : P.rels) :
+    P.jacobiMatrix i j = MvPolynomial.pderiv (P.map i) (P.relation j) := by
+  simp [jacobiMatrix, LinearMap.toMatrix, differential]
+
+end Matrix
+
+end PreSubmersivePresentation
+
+/--
+A `PreSubmersivePresentation` is submersive if its jacobian is a unit in `S`
+and the presentation is finite.
+-/
+@[nolint checkUnivs]
+structure SubmersivePresentation extends PreSubmersivePresentation.{t, w} R S where
+  jacobian_isUnit : IsUnit toPreSubmersivePresentation.jacobian
+  isFinite : toPreSubmersivePresentation.IsFinite := by infer_instance
+
+attribute [instance] SubmersivePresentation.isFinite
+
+/--
+An `R`-algebra `S` is called standard smooth, if there
+exists a submersive presentation.
+-/
+class IsStandardSmooth : Prop where
+  out : Nonempty (SubmersivePresentation.{t, w} R S)
+
+/--
+The relative dimension of a standard smooth `R`-algebra `S` is
+the dimension of an arbitrarily chosen submersive `R`-presentation of `S`.
+
+Note: This number is independent of the choice of the presentation as it is equal to
+the `S`-rank of `Ω[S/R]` (TODO).
+-/
+noncomputable def IsStandardSmooth.relativeDimension [IsStandardSmooth R S] : ℕ :=
+  ‹IsStandardSmooth R S›.out.some.dimension
+
+end Algebra
+
+namespace RingHom
+
+variable {R S}
+
+/-- A ring homomorphism `R →+* S` is standard smooth if `S` is standard smooth as `S`-algebra. -/
+def IsStandardSmooth (f : R →+* S) : Prop :=
+  @Algebra.IsStandardSmooth.{t, w} _ _ _ _ f.toAlgebra
+
+/-- A ring homomorphism `R →+* S` is standard smooth of relative dimension `n` if
+it is both standard smooth and is of relative dimension `n`. -/
+structure IsStandardSmoothOfRelativeDimension (n : ℕ) (f : R →+* S) : Prop where
+  isStandardSmooth : IsStandardSmooth.{t, w} f
+  relativeDimension_eq :
+    @Algebra.IsStandardSmooth.relativeDimension R _ S _ f.toAlgebra isStandardSmooth = n
```

Comment:
Here, I would rather say that there exists a submersive presentation that is of relative dimension `n`. Otherwise, you will have issues when proving that that if `f` is std smooth of relative dimension `a` and `g` of rel. dim.`b`, then `IsStandardSmooth (g.comp f) (a + b)`. The (silly) reason is that over the zero ring, the uniqueness of the rel. dim. is not true. The uniqueness shall hold only under the `Nontrivial S` assumption.